### PR TITLE
feat(workspaces): assign project roles to new users in a workspace

### PR DESCRIPTION
**Pull request by cdriesler**
## Description & motivation

- We are continuing to use project/stream roles for authorization purposes for projects in a workspace
- These roles are implied by, and need to be kept in sync with, a user's workspace role

TODO: It is unclear if project roles should always be kept in sync with workspace roles. Can I own a stream I create in a workspace without being an admin? Or maybe only admins can create projects? What happens if they stop being an admin? etc etc

## Changes:

- Adds logic to the `setWorkspaceRole` service function to update workspace project roles on change
- Adds logic to the `deleteWorkspaceRole` service function to delete workspace project roles

## Checklist:

- [ ] My pull request follows the guidelines in the [Contributing guide](https://github.com/specklesystems/speckle-server/blob/main/CONTRIBUTING.md)?
- [ ] My pull request does not duplicate any other open [Pull Requests](../../pulls) for the same update/change?
- [ ] My commits are related to the pull request and do not amend unrelated code or documentation.
- [ ] My code follows a similar style to existing code.
- [ ] I have added appropriate tests.
- [ ] I have updated or added relevant documentation.

### CODE REVIEW

**Review by gjedlicska — CHANGES_REQUESTED** [review]

**Comment by cdriesler** [comment]
@gjedlicska done and back to you

File: packages/server/modules/workspaces/repositories/workspaces.ts [hunk 1]
```diff
@@ -102,3 +105,11 @@ export const upsertWorkspaceRoleFactory =
       .onConflict(['userId', 'workspaceId'])
       .merge(['role'])
   }
+
+export const getWorkspaceProjectsFactory =
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/repositories/workspaces.ts:109]: Not to be that guy, but this doesn't seam to have integration tests

File: packages/server/modules/workspaces/utils/mapWorkspaceRoleToProjectRole.ts
Comment (gjedlicska) [inline packages/server/modules/workspaces/utils/mapWorkspaceRoleToProjectRole.ts:1]: This, arguably could be be `packages/server/modules/workspaces/DOMAIN/mapWorkspaceRoleToProjectRole.ts`

File: packages/server/modules/workspaces/repositories/workspaces.ts (hunk 1 above)
Comment (gjedlicska) [reply packages/server/modules/workspaces/repositories/workspaces.ts:109]: Also, i think it would be probably better to reuse the existing project repository function by extending it with an optional `workspaceId` argument. The benefit there would be, that the project list functionality will be consistent both in and outside of a workspace.

**Review by gjedlicska — CHANGES_REQUESTED** [review]

File: packages/server/modules/workspaces/utils/mapWorkspaceRoleToProjectRole.ts
Comment (fabis94) [reply packages/server/modules/workspaces/utils/mapWorkspaceRoleToProjectRole.ts:1]: Also: Why such an uber specific module name? One function per util module seems overly granular. I'd just have `roles.ts` with various role related helpers

File: packages/server/modules/workspaces/utils/mapWorkspaceRoleToProjectRole.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { StreamRoles, WorkspaceRoles } from '@speckle/shared'
+
+/**
+ * Given a user's workspace role, return the role they should have for workspace projects.
+ */
+export const mapWorkspaceRoleToProjectRole = (
+  workspaceRole: WorkspaceRoles
+): StreamRoles => {
+  switch (workspaceRole) {
+    case 'workspace:guest':
```
Comment (fabis94) [inline packages/server/modules/workspaces/utils/mapWorkspaceRoleToProjectRole.ts:10]: avoid string literals, use the `Roles` const from @speckle/shared

Comment (cdriesler) [reply packages/server/modules/workspaces/utils/mapWorkspaceRoleToProjectRole.ts]: roger, done

File: packages/server/modules/workspaces/repositories/workspaces.ts (hunk 1 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/repositories/workspaces.ts]: no, fair. I like extending the existing repo functions, but I'm not thrilled by the two options (I found):

- `getDiscoverable` is too focused on _public_ projects
- `getUser` requires a user, which we can do here but may not always be the case

I gave it a swing, though. What do you think? If you're ok with this strategy I'll come back in and add the tests.

File: packages/server/modules/workspaces/utils/mapWorkspaceRoleToProjectRole.ts
Comment (cdriesler) [reply packages/server/modules/workspaces/utils/mapWorkspaceRoleToProjectRole.ts:1]: moved and consolidated/renamed. just a personal bias for one-function-one-file when I'm making "utils" but no strong opinion in this case. `roles.ts` looks quite nice with the other `domain` files

File: packages/server/modules/workspaces/services/workspaceProjects.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { StreamRecord } from '@/modules/core/helpers/types'
```
Comment (fabis94) [inline packages/server/modules/workspaces/services/workspaceProjects.ts:1]: Do `workspaceProjects` and `workspaceProjectRoleCreation` really need to be 2 different services? It's kind of subjective, so if you think they do, then we can leave it at that, but I'd like to avoid having dozens of tiny services that only have 1 function in them

File: packages/server/modules/workspaces/services/workspaceProjects.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    do {
+      const { streams: workspaceProjects, cursor: pageCursor } = await getUserStreams({
+        userId,
+        workspaceIdWhitelist: [workspaceId],
+        cursor
+      })
+      projects.push(...workspaceProjects)
+      cursor = pageCursor ?? undefined
+    } while (!!cursor)
```
Comment (fabis94) [inline packages/server/modules/workspaces/services/workspaceProjects.ts:28]: 1. while/do-while is generally scary and should always have some kind of failsafe, e.g. always breaking after X loops. X can be something practically unreachable like 500 pages of projects, but at the very least it ends somewhere instead of executing indefinitely
2. just infinitely batching up all projects as long as there are more is likely to be a performance bomb - there could be a shit ton of projects OR there could be a bug in the repo query causing an infinitely increasing array of projects. instead, use JS generators and yield each page individually so that we never have to keep more than 1 page in memory - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator

File: packages/server/modules/workspaces/tests/unit/services/workspaceProjectRoleCreation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import { StreamAclRecord, StreamRecord } from '@/modules/core/helpers/types'
+import { WorkspaceAcl } from '@/modules/workspacesCore/domain/types'
+import { grantWorkspaceProjectRolesFactory } from '@/modules/workspaces/services/workspaceProjectRoleCreation'
+import { Roles } from '@speckle/shared'
+import { expect } from 'chai'
+import cryptoRandomString from 'crypto-random-string'
+
+describe('Workspace project role services', () => {
+  describe('grantWorkspaceProjectRolesFactory creates a function, that', () => {
```
Comment (fabis94) [inline packages/server/modules/workspaces/tests/unit/services/workspaceProjectRoleCreation.spec.ts:9]: Do we really need the test files to be this granular? There is an overheard to each test file - initializing a new test context, running `beforeEachContext` etc.

Some if not all of these can probably be merged into the same test file

Comment (fabis94) [reply packages/server/modules/workspaces/tests/unit/services/workspaceProjectRoleCreation.spec.ts:9]: Ah, this is a unit test, so its probably lightweight enough to stand on its own. That said, I'd still evaluate if some of the tests can't be merged just to avoid the overheard there is and also be able to reuse more of test setup, without having to duplicate it

**Review by fabis94 — CHANGES_REQUESTED** [review]

File: packages/server/modules/workspaces/tests/unit/services/workspaceProjectRoleCreation.spec.ts (hunk 5 above)
Comment (gjedlicska) [reply packages/server/modules/workspaces/tests/unit/services/workspaceProjectRoleCreation.spec.ts:9]: Its a unittest, being verbose in the test context is actually good.

File: packages/server/modules/workspaces/events/eventListener.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  const grantWorkspaceProjectRoles = grantWorkspaceProjectRolesFactory({
+    getWorkspaceRoles: getWorkspaceRolesFactory({ db }),
+    // TODO: Instantiate via factory function
+    grantStreamPermissions
+  })
+
+  await grantWorkspaceProjectRoles({ projectId, workspaceId })
+}
+
+export function initializeEventListener() {
```
[the remark this answers is not in the document]
Comment (cdriesler) [reply packages/server/modules/workspaces/events/eventListener.ts]: is better, done

File: packages/server/modules/workspaces/tests/unit/services/workspaceProjectRoleCreation.spec.ts (hunk 5 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/tests/unit/services/workspaceProjectRoleCreation.spec.ts]: will push back on this one - I think the verbosity/overhead is worth it here, but will keep this in mind going forwards

**Comment by cdriesler** [comment]
@fabis94 Back to you. No strong opinions on module names/boundaries here, just a personal default bias for smaller and more numerous files. I think this is now closer to what you're suggesting and to the kind of thing I see in other modules.

File: packages/server/modules/workspaces/services/management.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+    // Delete workspace project roles
+    const projectIds: string[] = []
+    for await (const project of queryAllWorkspaceProjectsFactory({ getStreams })(
+      workspaceId
+    )) {
+      projectIds.push(project.id)
+    }
+    await Promise.all(
+      projectIds.map((streamId) => revokeStreamPermissions({ streamId, userId }))
```
Comment (fabis94) [inline packages/server/modules/workspaces/services/management.ts:116]: I mean this works, but ideally we'd want the generator to return the projects in batches, instead of one by one, so that you can then also `revokeStreamPermissions` in batches. As opposed to now collecting potentially hundreds or thousands of IDs and then invoking hundreds/thousands of concurrent promises that may kill the server.

```ts
for await (const projectsPage of projectPageGenerator) {
  await Promise.all(projectsPage.map(p => revokeStreamPermissions({ streamId: p.id, userId })))
}
```

That said, this is already better than what was here before

File: packages/server/modules/workspaces/services/management.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    )
+
+    // Emit deleted role
+    emitWorkspaceEvent({ eventName: WorkspaceEvents.RoleDeleted, payload: deletedRole })
```
Comment (fabis94) [inline packages/server/modules/workspaces/services/management.ts:120]: Isn't this async? You should await these

File: packages/server/modules/workspaces/services/projects.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        orderBy: null,
+        limit: 1000,
+        visibility: null,
+        searchQuery: null,
+        streamIdWhitelist: null,
+        workspaceIdWhitelist: [workspaceId]
+      })
+
+      for (const stream of streams ?? []) {
+        yield stream
```
Comment (fabis94) [inline packages/server/modules/workspaces/services/projects.ts:31]: as mentioned above, i'd yield the entire batch so you can work with it in a more optimized way

File: packages/server/modules/workspaces/tests/unit/services/management.spec.ts [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+        role: Roles.Workspace.Member
+      }
+      const workspaceRoles: WorkspaceAcl[] = [workspaceRole]
+      const workspaceProjects: StreamRecord[] = [{ id: projectId } as StreamRecord]
+
+      let projectRoles: StreamAclRecord[] = [
+        { userId, role: Roles.Stream.Contributor, resourceId: projectId }
+      ]
+
+      const deleteWorkspaceRole = deleteWorkspaceRoleFactory({
```
Comment (fabis94) [inline packages/server/modules/workspaces/tests/unit/services/management.spec.ts:214]: Figuring out how DRY tests should be is an art form in and of itself - if you go too DRY you can run into way too abstract tests that are too hard to understand and change (e.g. `modules/comments/tests/comments.graph.spec.js`). That said, copy and pasting the same setup over and over again when it's expected to be the same for most if not all test cases, even if it does change in the future, is going too far in the other direction - just changing the deps a little bit for the role factory will require the dev to go and make a dozen identical changes to all factory invocations across this test file.

I definitely want to push back about this assumption that DRY is not a concern in tests - it most definitely is, you just should not go overboard with it and make it "too clever" and you have to evaluate which bits of code are duplicate only currently and very likely to differ in the future and which are more or less always going to be the same. In this case there's so much duplication of common init, it could easily be extracted to a helper at the top of the file. In the case of factory init, you can also always provide a param that allows you to override some of the dependencies, if needed, while all of the other ones stay standard.

For examples on how I've approached this you can check `automations.spec.ts` or `trigger.spec.ts`

**Review by fabis94 — CHANGES_REQUESTED** [review]

File: packages/server/modules/workspaces/services/management.ts (hunk 7 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/services/management.ts]: I went back and forth but this is a very good reason to go by page, done

File: packages/server/modules/workspaces/services/management.ts (hunk 8 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/services/management.ts]: ope, yes

File: packages/server/modules/workspaces/tests/unit/services/management.spec.ts (hunk 10 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/tests/unit/services/management.spec.ts]: (discussed in person but some late thoughts) I think this has landed somewhere much better, lmk what you think, and ty for the push

**Review by fabis94 — APPROVED** [review]

**Review by gjedlicska — DISMISSED** [review]

**Review by gjedlicska — APPROVED** [review]
